ev3dev2/console.py

```diff
@@ -0,0 +1,104 @@
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+# -----------------------------------------------------------------------------
+
+import os
+
+
+class Console():
+    """
+    A class that represents the EV3 LCD console, which implements ANSI codes
+    for cursor positioning, text color, and clearing the screen. Supports changing
+    the console font using standard system fonts.
+    """
+
+    def __init__(self, font="Lat15-TerminusBold24x12"):
+        """
+        Construct the Console instance, optionally with a font name specified.
+        This also resets the console: clearing it and turning off the cursor.
+        Parameter:
+
+        - `font` (string): Font name, as found in `/usr/share/consolefonts/`
+
+        """
+        self._font = None
+        self.set_font(font)
+
+    def text_at(self, text, column=1, row=1, clear_screen=False, inverse=False):
+        """
+        Display 'text' (string) starting at grid (column, row).
+        Note that the grid locations are 1-based (not 0-based).
+
+        Depending on the font, the EV3 LCD console can show text within 4 to 21 rows, and 11 to 44 columns.
+        The default font results in a grid that is 14 columns wide and 5 rows tall.
+
+        Parameters:
+
+        - `text` (string): Text to display
+        - `column` (int): LCD column position to start the text (1 = left column);
+          text will wrap when it reaches the right edge
+        - `row` (int): LCD row position to start the text (1 = top row)
+        - `clear_screen` (bool): ``True`` to clear the screen before showing the text; default is ``False``
+        - `inverse` (bool): ``True`` for white, otherwise black; default is ``False``
+
+        """
+
+        if clear_screen:
+            self.reset_screen()
+
+        if inverse:
+            text = "\x1b[7m%s\x1b[m" % (text)
+
+        print("\x1b[%d;%dH%s" % (row, column, text), end='')
+
+    def set_font(self, font):
+        """
+        Set the LCD console font, clear the screen and turn off the cursor.
+        Parameter:
+
+        - `font` (string): Font name, as found in `/usr/share/consolefonts/`
+
+        """
+        if font is not None and font != self._font:
+            self._font = font
+            os.system("setfont %s" % (font))
+        self.reset_screen()

```

Comment:
Can we avoid the `reset_screen` so multiple fonts can be displayed on the screen?

Reply:
I just verified... the `setfont` OS command changes the font for the whole EV3 LCD console at once--i.e. it doesn't enable showing text with two or more different fonts on screen at the same time. But your correction would allow to change the font while text was already on the screen--enabling some possible text-based presto font-change-o animations... Let me re-tool this function to accept a `clear_screen` parameter (default True, as most often folks will want to clear the screen after a font change).